staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/clock/event_clock.go

```diff
@@ -0,0 +1,251 @@
+	ch := time.After(d)
+	go func() {
+		select {
+		case t := <-ch:
+			f(t)
+		}
+	}()
+}
+
+// EventAfterTime schedules an EventFunc
+func (r RealEventClock) EventAfterTime(f EventFunc, t time.Time) {
+	now := time.Now()
+	d := t.Sub(now)
+	if d <= 0 {
+		go f(now)
+	} else {
+		r.EventAfterDuration(f, d)
+	}
+}
+
+// waitGroupCounter is a wait group used for a GoRoutine Counter.  This private
+// type is used to disallow direct waitGroup access
+type waitGroupCounter struct {
+	wg sync.WaitGroup
+}
+
+// compile time assertion that waitGroupCounter meets requirements
+//  of GoRoutineCounter
+var _ counter.GoRoutineCounter = (*waitGroupCounter)(nil)
+
+func (wgc *waitGroupCounter) Add(delta int) {
+	if klog.V(7) {
+		var pcs [5]uintptr
+		nCallers := runtime.Callers(2, pcs[:])
+		frames := runtime.CallersFrames(pcs[:nCallers])
+		frame1, more1 := frames.Next()
+		fileParts1 := strings.Split(frame1.File, "/")
+		tail2 := "(none)"
+		line2 := 0
+		if more1 {
+			frame2, _ := frames.Next()
+			fileParts2 := strings.Split(frame2.File, "/")
+			tail2 = fileParts2[len(fileParts2)-1]
+			line2 = frame2.Line
+		}
+		klog.Infof("GRC(%p).Add(%d) from %s:%d from %s:%d", wgc, delta, fileParts1[len(fileParts1)-1], frame1.Line, tail2, line2)
+	}
+	wgc.wg.Add(delta)
+}
+
+func (wgc *waitGroupCounter) Wait() {
+	wgc.wg.Wait()
+}
+
+// FakeEventClock is one whose time does not pass implicitly but
+// rather is explicitly set by invocations of its SetTime method
+type FakeEventClock struct {
+	clock.FakePassiveClock
+
+	// waiters is a heap of waiting work, sorted by time
+	waiters     eventWaiterHeap
+	waitersLock sync.RWMutex
+
+	// clientWG may be nil and if not supplies constraints on time
+	// passing in Run.  The Run method will not pick a new time until
+	// this is nil or its counter is zero.
+	clientWG *waitGroupCounter
+
+	// fuzz is the amount of noise to add to scheduling.  An event
+	// requested to run at time T will run at some time chosen
+	// uniformly at random from the interval [T, T+fuzz]; the upper
+	// bound is exclusive iff fuzz is non-zero.
+	fuzz time.Duration
+
+	// rand is the random number generator to use in fuzzing
+	rand *rand.Rand
+}
+
+type eventWaiterHeap []eventWaiter
+
+var _ heap.Interface = (*eventWaiterHeap)(nil)
+
+type eventWaiter struct {
+	targetTime time.Time
+	f          EventFunc
+}
+
+// NewFakeEventClock constructor.  The given `r *rand.Rand` must
+// henceforth not be used for any other purpose.  If `r` is nil then a
+// fresh one will be constructed, seeded with the current real time.
+// The clientWG can be `nil` and if not is used to let Run know about
+// additional work that has to complete before time can advance.
+func NewFakeEventClock(t time.Time, fuzz time.Duration, r *rand.Rand) (*FakeEventClock, counter.GoRoutineCounter) {
+	grc := &waitGroupCounter{}
+
+	if r == nil {
+		r = rand.New(rand.NewSource(time.Now().UnixNano()))
+		r.Uint64()
```

Comment:
Why calling Uint64 3 times ?
It seems the calls can be removed.

Reply:
My experience in a previous life was that the constructor of a random generator did not do a really good job of scrambling the bits from the seed; generating a few random numbers did the trick.  I am not sure whether that's needed here and now.

Reply:
I have never seen such pattern in k8s w.r.t. the usage of random.
I think the extra calls are not needed.